src/renderers/shared/shared/event/ReactGenericBatching.js

```diff
@@ -36,6 +38,29 @@ function batchedUpdates(fn, bookkeeping) {
   stackBatchedUpdates(performFiberBatchedUpdates, fn, bookkeeping);
 }
 
+var isBatching = false;
+function batchedUpdatesWithControlledTarget(fn, bookkeeping, target) {
+  if (isBatching) {
+    // TODO: If this target is not the same as the one currently batched,
+    // we'll drop it.
```

Comment:
I think that this is already a problem with the current style of batching. If a controlled event happens *within* another batch, we won't flush completely down until after the controlled event is already done.